src/named_tuple_dist.jl

```diff
@@ -144,7 +144,7 @@ end
 function _ntd_logpdf(d::NamedTupleDist, x::AbstractVector{<:Real})
     distributions = values(d)
     accessors = values(varshape(d))
-    sum(map((dist, acc) -> _ntd_logpdf(dist, acc, x), distributions, accessors))
+    mapreduce((dist, acc) -> _ntd_logpdf(dist, acc, x), +, distributions, accessors)
```

Comment:
I don't think this is faster than `sum(map(...`, and it's not type stable if the return type of `logpdf` varies. For example, with

```julia
using ValueShapes, Distributions
using ValueShapes: _ntd_logpdf
d = NamedTupleDist(a = Normal(), b = Normal(0.0f0, 1.0f0), c = Exponential(1.0f0))
x = Float32[1, 1, 1]

distributions = values(d)
accessors = values(varshape(d))
```

`sum(map(...))` is type stable, but `mapreduce(...)` is not:

```
julia> @code_warntype sum(map((dist, acc) -> _ntd_logpdf(dist, acc, x), distributions, accessors))
[...]
Body::Float64
[...]

julia> @code_warntype mapreduce((dist, acc) -> _ntd_logpdf(dist, acc, x), +, distributions, accessors)
[...]
Body::Any
[...]
```

And

```julia
julia> @code_typed map((dist, acc) -> _ntd_logpdf(dist, acc, x), distributions, accessors)
CodeInfo(
1 ─ %1  = Base.getfield(t, 1, true)::Normal{Float64}
│   %2  = Base.getfield(s, 1, true)::ValueAccessor{ScalarShape{Real}}
│   %3  = invoke f(%1::Normal{Float64}, %2::ValueAccessor{ScalarShape{Real}})::Any
│   %4  = Core.getfield(t, 2)::Normal{Float32}
│   %5  = Core.getfield(t, 3)::Exponential{Float32}
│   %6  = Core.getfield(s, 2)::ValueAccessor{ScalarShape{Real}}
│   %7  = Core.getfield(s, 3)::ValueAccessor{ScalarShape{Real}}
│   %8  = invoke f(%4::Normal{Float32}, %6::ValueAccessor{ScalarShape{Real}})::Any
│   %9  = invoke f(%5::Exponential{Float32}, %7::ValueAccessor{ScalarShape{Real}})::Any
│   %10 = Core.tuple(%3, %8, %9)::Tuple{Any, Any, Any}
└──       return %10
) => Tuple{Any, Any, Any}

julia> @code_typed sum(map((dist, acc) -> _ntd_logpdf(dist, acc, x), distributions, accessors))
CodeInfo(
1 ─ %1 = Core.getfield(x, 1)::Float64
│   %2 = Core.getfield(x, 2)::Float32
│   %3 = Core.getfield(x, 3)::Float32
│   %4 = Base.fpext(Base.Float64, %2)::Float64
│   %5 = Base.add_float(%1, %4)::Float64
│   %6 = Base.fpext(Base.Float64, %3)::Float64
│   %7 = Base.add_float(%5, %6)::Float64
└──      return %7
) => Float64

julia> @benchmark sum(map((dist, acc) -> _ntd_logpdf(dist, acc, $x), $distributions, $accessors))
BenchmarkTools.Trial: 10000 samples with 997 evaluations.
 Range (min … max):  19.555 ns … 212.825 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     22.628 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   23.016 ns ±   3.507 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%
```

shows that `sum(map(...))` generates compact (and allocation-free) code.

Reply:
That's weird because map must allocate an intermediate array first.

Also I would expect the same type instability to exist in map if it's from logpdf.

In my use case mapreduce is also type stable, you might want to check the pdf functions used here?

Reply:
> That's weird because map must allocate an intermediate array first.

No, it returns a tuple, not an array, and `sum` and `map` for tuples are highly optimized.

> In my use case mapreduce is also type stable

It can be type stable if the return type of `logpdf` is the same for all pairs of entries. But if it's not, e.g. `Float32` for some and `Float64` for others (like in the example above), it's not type stable because `mapreduce` "reuses" the accumulation variable.